lib/puppet/resource_api/value_creator.rb

```diff
@@ -0,0 +1,87 @@
+module Puppet; module ResourceApi; end; end # predeclare the main module # rubocop:disable Style/Documentation,Style/ClassAndModuleChildren
+
+# This class is responsible for setting default and alias values for the
+# resource class.
+class Puppet::ResourceApi::ValueCreator
+  # This initialize takes arguments and sets up new value creator for given
+  # resource class which can be Puppet::ResourceApi::Parameter,
+  # Puppet::ResourceApi::ReadOnlyParameter, Puppet::ResourceApi::Property.
+  # @param resource_class the class of selected resource to be extended
+  # @param data_type the type instance
+  # @param definition the definition of the property
+  # @param options the ResourceAPI options hash containing setup information for
+  # selected parameter or property
+  def initialize(resource_class, data_type, param_or_property, options = {})
+    @resource_class = resource_class
+    @data_type = data_type
+    @param_or_property = param_or_property
+    @options = options
+  end
+
+  # This method is responsible for all setup of the value mapping for desired
+  # resource class.
+  def create_values
+    @resource_class.isnamevar if @options[:behaviour] == :namevar
+
+    # read-only values do not need type checking, but can have default values
+    if @options[:behaviour] != :read_only && @options.key?(:default)
+      if @options[:default] == false
+        # work around https://tickets.puppetlabs.com/browse/PUP-2368
+        @resource_class.defaultto :false # rubocop:disable Lint/BooleanSymbol
+      elsif @options[:default] == true
+        # work around https://tickets.puppetlabs.com/browse/PUP-2368
+        @resource_class.defaultto :true # rubocop:disable Lint/BooleanSymbol
+      else
+        # marshal the default option to decouple that from the actual value.
+        # we cache the dumped value in `marshalled`, but use a block to
+        # unmarshal everytime the value is requested. Objects that can't be
+        # marshalled
+        # See https://stackoverflow.com/a/8206537/4918
+        marshalled = Marshal.dump(@options[:default])
+        @resource_class.defaultto { Marshal.load(marshalled) } # rubocop:disable Security/MarshalLoad
+      end
+    end
+
+    case @data_type
```

Comment:
This is the place where the `Optional` unpacking from https://github.com/puppetlabs/puppet-resource_api/blob/9dd87a93c8e0926a90745058df5005d311825426/lib/puppet/resource_api.rb#L299-L302 should go. `@data_type` must always be the full unmodified type from `parse_puppet_type` to avoid future confusion.